File: src/main/java/dev/esophose/discordbot/command/arguments/SnowflakeArgumentHandler.java
package dev.esophose.discordbot.command.arguments;

import dev.esophose.discordbot.command.DiscordCommandArgumentHandler;
import discord4j.core.object.entity.Guild;
import discord4j.core.object.util.Snowflake;
import reactor.core.publisher.Mono;

public class SnowflakeArgumentHandler extends DiscordCommandArgumentHandler<Snowflake> {

    @Override
    protected Mono<Snowflake> handleInternal(Guild guild, String input) {
        try {
            long id = Long.parseUnsignedLong(input);
            return Mono.just(Snowflake.of(id));
        } catch (Exception ex) {
            return Mono.empty();
        }
    }

    @Override
    public String getErrorMessage(Guild guild, String input) {
        return "Invalid snowflake: [" + input + "]";
    }

    @Override
    public Class<Snowflake> getHandledType() {
        return Snowflake.class;
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/arguments/SnowflakeArgumentHandler.kt
package dev.esophose.discordbot.command.arguments

import dev.esophose.discordbot.command.DiscordCommandArgumentHandler
import discord4j.common.util.Snowflake
import discord4j.core.`object`.entity.Guild
import reactor.core.publisher.Mono
import kotlin.reflect.KClass

class SnowflakeArgumentHandler : DiscordCommandArgumentHandler<Snowflake>() {

    override val handledType: KClass<Snowflake>
        get() = Snowflake::class

    public override fun handleInternal(guild: Guild, input: String): Mono<Snowflake> {
        return try {
            val id = java.lang.Long.parseUnsignedLong(input)
            Mono.just(Snowflake.of(id))
        } catch (ex: Exception) {
            Mono.empty()
        }
    }

    override fun getErrorMessage(guild: Guild, input: String): String {
        return "Invalid snowflake: [$input]"
    }

}
